ffi.el

```diff
@@ -1,21 +1,40 @@
-;; -*- lexical-binding:t -*-
+;;; ffi.el --- -*- lexical-binding: t -*-
+
+;; Copyright (C) 2015-2017 Tom Tromey
+
+;; Author: Tom Tromey <tromey@redhat.com>
+;; Package-Requires: ((emacs "25.1"))
+
+;; TODO Put permission statement here and also add LICENSE file.
+
+;;; Commentary:
+
+;; This is an FFI for Emacs.  It is based on libffi and relies on the
+;; dynamic module support in order to be loaded into Emacs.  It is
+;; relatively full-featured, but for the time being low-level.
+
+;;; Code:
 
 (require 'cl-macs)
 
-(module-load "ffi-module.so")
+(defun ffi--locate-module (name &optional nosuffix path)
+  (let ((load-suffixes (list module-file-suffix)))
+    (locate-library name)))
+
+(module-load (ffi--locate-module "ffi-module"))
 
 (gv-define-simple-setter ffi--mem-ref ffi--mem-set t)
 
-(defmacro define-ffi-library (symbol name)
+(defmacro define-ffi-library (symbol name &optional filename)
   (let ((library (cl-gensym)))
     (set library nil)
     `(defun ,symbol ()
        (or ,library
-	   (setq ,library (ffi--dlopen ,name))))))
+	   (setq ,library
+		 (ffi--dlopen (or ,filename (ffi--locate-module ,name))))))))
```

Comment:
I saw your note about `ffi-library-path` and I also wonder if that would be better.  Also it's probably important to allow loading of libraries from the system library paths -- that is, let the user pass in a relative path and hope that `dlopen` finds it.